src/expander/expander.c

```diff
@@ -0,0 +1,127 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   expander.c                                         :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: alopez-g <marvin@42.fr>                    +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2022/10/16 10:05:08 by alopez-g          #+#    #+#             */
+/*   Updated: 2022/11/01 23:32:10 by r3dc4t-g         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "expander.h"
+#include "reader.h"
+#include "tab.h"
+
+// TODO CHECK HUGE AMMOUNT OF LEAKS
+// TODO EXPAND $?
+// TODO IMPLEMENT IT TO MAIN FLOW AND USE QUOTE REMOVAL
+
+// CALL EXPANDER JUST AFTER READLINE (INSIDE READER_GET_TAB)
+// EXPANDER SHOULD RETURN EVERYTHING AS A WORD (BETWEEN QUOTES -> "word")
+static char	*str_append_char(char **str, char c, t_bool sufix)
+{
+	char	*ret;
+	char	*app;
+
+	app = (char *)malloc(sizeof(char) * 2);
+	*app = c;
+	*(app + 1) = 0;
+	if (sufix)
+		ret = ft_strjoin(*str, app);
+	else
+		ret = ft_strjoin(app, *str);
+	free(*str);
+	free(app);
+	return (ret);
+}
+
+static char	*word_encloser(char *str)
+{
+	char	*final_ret;
+	char	*tmp;
+	char	**ret;
+	size_t	i;
+
+	i = 0;
+	final_ret = ft_strdup("");
+	reader_split_by_token(str, &ret);
+	while (*(ret + i))
+	{
+		*(ret + i) = str_append_char((ret + i), '\"', FALSE);
+		*(ret + i) = str_append_char((ret + i), '\"', TRUE);
+		if (*(ret + i + 1))
+			*(ret + i) = str_append_char((ret + i), ' ', TRUE);
+		tmp = final_ret;
+		final_ret = ft_strjoin(final_ret, *(ret + i));
+		free(tmp);
+		i++;
+	}
+	tab_deep_destroy(&ret);
+	return (final_ret);
+}
+
+static int	get_error_var(t_shell_op sp, char **ret)
```

Comment:
it is better to pass objects as pointers since the obj can be very big and slow the stack call operation